lib/mirage/mirage.mli

```diff
@@ -764,6 +764,114 @@ val default_argv : argv impl
 val no_argv : argv impl
 (** [no_argv] Disable command line parsing and set argv to [|""|]. *)
 
+(** {2 Git client configuration} *)
+
+(** The user is able to initiate a connection with a Git repository. However,
+    there are many ways to communicate with a Git repository:
+
+    - TCP/IP
+    - HTTP
+    - HTTP + TLS
+    - SSH
+
+    The devices defined below define these composable ways. The
+    [git_client impl] returned from them can be passed to Git or Irmin in order
+    to be able to {i fetch} and {i push} from/into a Git repository.
+
+    The user is able to restrict or enlarge protocol possibilities needed for
+    its application. For instance, the user is able to restrict only the SSH
+    connection to communicate with a Git repository or the user can handle
+    TCP/IP and SSH as possible protocols to communicate with a peer.
+
+    For instance, a device which is able to communicate {i via} TCP/IP and SSH
+    can be implemented like:
+
+    {[
+      let git_client =
+        let dns = git_happy_eyeballs stackv4v6 in
+        let ssh = git_ssh ~key (git_tcpv4v6_of_stackv4v6 stackv4v6) dns in
+        let tcp = git_tcp (git_tcpv4v6_of_stackv4v6 stackv4v6) dns in
+        git_merge_clients ssh tcp
+    ]} *)
+
+type git_client
+(** Type of a device which implements a {i certain} protocol. *)
+
+val git_client : git_client typ
+
+type git_client_with_dns
+(** Type of a device which is able to {b only} resolve domain-name. *)
+
+val git_client_with_dns : git_client_with_dns typ
+
+val git_merge_clients : git_client impl -> git_client impl -> git_client impl
+(** [git_merge_clients a0 a1] merges given [git_client] into one. *)
+
+val git_happy_eyeballs :
+  ?random:random impl ->
+  ?time:time impl ->
+  ?mclock:mclock impl ->
+  ?pclock:pclock impl ->
+  stackv4v6 impl ->
+  git_client_with_dns impl
+(** [git_happy_eyeballs stackv4v6] provides a {!git_client_with_dns} device *
+    which is {b only} able to resolve domain-name. Then, the user is able to
+    create a new device {!git_client} from it which will provide a {i certain}
+    protocol. The resolution of domain-name is done {i via} a {i happy-eyeballs}
+    query. *)
+
+val git_tcp : tcpv4v6 impl -> git_client_with_dns impl -> git_client impl
+(** [git_tcp tcpv4v6 ctx] returns a new {!mimic} context which is able to
+    allocate a TCP/IP connection with a remote Git repository. *)
+
+val git_ssh :
+  ?authenticator:string option key ->
+  key:string option key ->
+  ?mclock:mclock impl ->
+  ?time:time impl ->
+  tcpv4v6 impl ->
+  git_client_with_dns impl ->
+  git_client impl
+(** [git_ssh ?authenticator ~key tcpv4v6 ctx] returns a new {!mimic} context
+    which is able to allocate an SSH connection with the given private [key]
+    with a remote Git repository. You can ensure the identity of the remote Git
+    repository via [authenticator].
```

Comment:
```suggestion
(** [git_ssh ?authenticator ~key tcpv4v6 dns] is device able to connect to a
   remote Git repository using an SSH connection with the given private [key].
    You can ensure the identity of the remote Git repository via [authenticator].
```

Reply:
What should the string provided to `authenticator` contain? Is it the name of a directory in a KV-RO store?